blog: introducing kube-scheduler-simulator (#47104)
kubernetes/website: content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md

---
layout: blog

> title: "Introducing kube-scheduler-simulator"

Reviewer: We should get an answer about which way you'd like to go on this one @sanposhiho. It seems important we're using the right name that end users would typically know the software by.

date: 2025-12-31
draft: true 
slug: introducing-kube-scheduler-simulator
author: Kensei Nakada (Tetrate)
---

> The Kubernetes Scheduler is a crucial control plane component that determines which node a Pod will run on. 

Reviewer 2: ```suggestion
The [Kubernetes Scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/) is a crucial control plane component that determines which node a Pod will run on. 
```

Reviewer: (@sanposhiho you can make that change, but it's not a blocker for publication)

Thus, anyone utilizing Kubernetes relies on a scheduler.

The [kube-scheduler-simulator](https://sigs.k8s.io/kube-scheduler-simulator) is a simulator for the Kubernetes scheduler, started as a [Google Summer of Code 2021](https://summerofcode.withgoogle.com/) project developed by me (Kensei Nakada) and later received a lot of contributions.
This tool allows users to closely examine the scheduler’s behavior and decisions. 

> It is useful for casual users who employ scheduling constraints (e.g., [inter-Pod affinity](/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity))

Reviewer: (@sanposhiho you can make that change, but it's not a blocker for publication)

and experts who extend the scheduler with custom plugins.

## Motivation

The scheduler often appears as a black box, 
composed of many plugins that each contribute to the scheduling decision-making process from their unique perspectives. 
Understanding its behavior can be challenging due to the multitude of factors it considers. 
Even if a Pod seems to be scheduled as expected in a simple test cluster, 
it may be coming from a different calculation than the expectation, 

> which could result in unexpected scheduling results in a large production environment.

Reviewer 2: > Even if a Pod seems to be scheduled as expected in a simple test cluster, 
it may be coming from a different calculation than the expectation, 
which could result in unexpected scheduling results in a large production environment.

Try to make it like:

> Even if a Pod appears to be scheduled correctly in a simple test cluster, it might have been scheduled based on different calculations than expected. This discrepancy could lead to unexpected scheduling outcomes when deployed in a large production environment.

It's more readable.

Also, testing a scheduler is a complex challenge.

> There are countless patterns of operations executed within a real cluster, making it impractical to anticipate every scenario with a finite number of tests. 

Reviewer 2: ```suggestion
There are countless patterns of operations executed within a real cluster, making it unfeasible to anticipate every scenario with a finite number of tests. 
```

Reviewer: (@sanposhiho you can make that change, but it's not a blocker for publication)

More often than not, bugs are discovered only when the scheduler is deployed in an actual cluster.
Actually, many bugs are found by users after shipping the release, 
even in the upstream kube-scheduler. 

Having a development or sandbox environment for testing the scheduler — or, indeed, any Kubernetes controllers — is a common practice.
However, this approach falls short of capturing all the potential scenarios that might arise in a production cluster 
because a development cluster is often much smaller with notable differences in workload sizes and scaling dynamics.
It never sees the exact same use or exhibits the same behavior as its production counterpart.

> kube-scheduler-simulator aims to solve those problems.

Reviewer 2: ```suggestion
The kube-scheduler-simulator aims to solve those problems.
```

Reviewer: (@sanposhiho you can make that change, but it's not a blocker for publication)

It enables users to test their scheduling constraints, scheduler configurations, 
and custom plugins while checking every detailed part of scheduling decisions.
It also allows users to create a simulated cluster environment, where they can test their scheduler
with the same resources as their production cluster without affecting actual workloads.

## Features of the kube-scheduler-simulator

> kube-scheduler-simulator’s core feature is its ability to expose the scheduler's internal decisions.

Reviewer 2: ```suggestion
The kube-scheduler-simulator’s core feature is its ability to expose the scheduler's internal decisions.
```

Reviewer: (@sanposhiho you can make that change, but it's not a blocker for publication)

The scheduler operates based on the [scheduling framework](/docs/concepts/scheduling-eviction/scheduling-framework/), 

> utilizing various plugins at different extension points,

Reviewer 2: ```suggestion
using various plugins at different extension points,
```

Reviewer: (@sanposhiho you can make that change, but it's not a blocker for publication)

filter nodes (Filter phase), score nodes (Score phase), and ultimately determine the best node for the Pod.

The simulator allows users to create Kubernetes resources and observe how each plugin influences the scheduling decisions for Pods.
This visibility helps users understand the scheduler’s workings and define appropriate scheduling constraints.

{{< figure src="/images/blog/2025-01-22-kube-scheduler-simulator/simulator.png" alt="Screenshot of the simulator web frontend that shows the detailed scheduling results per node and per extension point" title="The simulator web frontend" >}}

Inside the simulator, a debuggable scheduler runs instead of the vanilla scheduler. 

> This debuggable scheduler outputs the results of each scheduler plugin at every extension point to the Pod’s annotations like the following Yaml shows

Reviewer 2: ```suggestion
This debuggable scheduler outputs the results of each scheduler plugin at every extension point to the Pod’s annotations like the following manifest shows
```

Reviewer: (@sanposhiho you can make that change, and on this one I do recommend it, but it's not a blocker for publication)

            "kube-scheduler-simulator.sigs.k8s.io/selected-node":"node-mtb5x"
        }
      ]
    kube-scheduler-simulator.sigs.k8s.io/score-result: >-
      {
        "node-jjfg5":{
            "ImageLocality":"0",
            "NodeAffinity":"0",
            "NodeResourcesBalancedAllocation":"52",
            "NodeResourcesFit":"47",
            "TaintToleration":"0",
            "VolumeBinding":"0"
        },
        "node-mtb5x":{
            "ImageLocality":"0",
            "NodeAffinity":"0",
            "NodeResourcesBalancedAllocation":"76",
            "NodeResourcesFit":"73",
            "TaintToleration":"0",
            "VolumeBinding":"0"
        }
      }
    kube-scheduler-simulator.sigs.k8s.io/selected-node: node-mtb5x
```

Users can also integrate [their custom plugins](/docs/concepts/scheduling-eviction/scheduling-framework/) or [extenders](https://github.com/kubernetes/design-proposals-archive/blob/main/scheduling/scheduler_extender.md), into the debuggable scheduler and visualize their results. 

> This debuggable scheduler can also run standalone, e.g., on any Kubernetes cluster or in integration tests. 

Reviewer 2: The style guides suggest to avoid latin phrases like "e.g" or "i.e"

https://kubernetes.io/docs/contribute/style/style-guide/#avoid-latin-phrases
```suggestion
This debuggable scheduler can also run standalone, for example, on any Kubernetes cluster or in integration tests. 
```

Reviewer: (@sanposhiho you can make that change, but it's not a blocker for publication)

This would be useful to custom plugin developers who want to test their plugins or examine their custom scheduler in a real cluster with better debuggability.

## The simulator as a better dev cluster

As mentioned earlier, with a limited set of tests, it is impossible to predict every possible scenario in a real-world cluster.
Typically, users will test the scheduler in a small, development cluster before deploying it to production, hoping that no issues arise.

[The simulator’s importing feature](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/docs/import-cluster-resources.md)
provides a solution by allowing users to simulate deploying a new scheduler version in a production-like environment without impacting their live workloads.

By continuously syncing between a production cluster and the simulator, users can safely test a new scheduler version with the same resources their production cluster handles. 
Once confident in its performance, they can proceed with the production deployment, reducing the risk of unexpected issues.

## What are the use cases?

> 1. **Cluster users**: Examine if scheduling constraints (e.g., PodAffinity, PodTopologySpread) work as intended.

Reviewer 2: The style guides suggest to avoid latin phrases like "e.g" or "i.e"

https://kubernetes.io/docs/contribute/style/style-guide/#avoid-latin-phrases

```suggestion
1. **Cluster users**: Examine if scheduling constraints (for example PodAffinity, PodTopologySpread) work as intended.
```

Reviewer: (@sanposhiho you can make that change, but it's not a blocker for publication)
